File: day5/p2/solution.exs
# This is slow in Elixir, so I probably need to learn some
# tricks to make list traversal and recursion faster.

defmodule Day5 do
  def num_of_steps(instructions, steps \\ 0, curr_index \\ 0) do
    if curr_index < 0 || curr_index >= Enum.count(instructions) do
      steps
    else
      offset = Enum.at(instructions, curr_index)
      offset_modifier = if offset >= 3, do: -1, else: 1
      next_index = curr_index + offset
      num_of_steps(
        List.replace_at(instructions, curr_index, offset + offset_modifier),
        steps + 1,
        next_index
      )
    end
  end
end

{:ok, input} = File.read("input.txt")
instructions = input
|> String.split("\n", trim: true)
|> Enum.map(fn(num) -> String.to_integer(num) end)

IO.inspect(Day5.num_of_steps(instructions))

File: day5/p2/solution.rb
def num_of_steps(instructions)
  steps = curr_index = next_index = 0
  while curr_index >= 0 && curr_index < instructions.length
    next_index = curr_index + instructions[curr_index]
    instructions[curr_index] += instructions[curr_index] >= 3 ? -1 : 1
    curr_index = next_index
    steps += 1
  end
  steps
end


instructions = File.read('input.txt').split("\n").map(&:to_i)
p num_of_steps(instructions)
